### PR TITLE
Considered preferred text output styles for heuristics

### PR DESCRIPTION
**Pull request by doshi-kevin**
## References

 #16636 

## Code changes

1. Expanded Supported Output Types
Change:
The list of supported output types has been expanded to include additional MIME types that are commonly used in Jupyter notebooks, specifically:
application/vnd.jupyter.stderr
application/vnd.jupyter.stdout
text/html
image/svg+xml
application/pdf
text/markdown
text
Purpose:
This change ensures that the height estimation considers outputs from standard error and standard output streams, which may contain significant text that contributes to the overall height of the cell. By including these types, the estimation becomes more comprehensive and accurate.
2. Dynamic Output Type Selection
Change:
The code now dynamically selects the preferred output type from the expanded list of supported types using the find method. It checks if the data associated with each output type is a string, allowing for a fallback to the first valid type found.
Purpose:
This approach allows the function to prioritize output types that are likely to contain meaningful text data. By finding the first output type that yields a string, the code ensures that it captures the most relevant output for height estimation, rather than defaulting to a less informative type.
3. Improved Line Counting Logic
Change:
The logic for counting lines in the output data has been refined. The code now checks if the data to display is defined and then counts non-empty lines using a filter.
Purpose:
This enhancement ensures that only meaningful lines are counted, which directly impacts the accuracy of the height estimation. By filtering out empty lines, the function avoids inflating the estimated height with irrelevant data.
4. Total Height Calculation
Change:
The total height is calculated by summing the number of lines in the source code (nLines) and the calculated output lines (outputsLines). This total is then stored in the cellsEstimatedHeight map for future reference.
Purpose:
This final calculation provides a complete and accurate estimation of the widget size. By combining both the source code and relevant output lines, the function delivers a more realistic height for the cell, which is essential for proper rendering in the UI.
How These Changes Address the Issue
The primary issue was that the original height estimation heuristic only considered text/plain outputs, which did not account for other significant output types that could affect the visual representation of the cell. By expanding the supported output types and dynamically selecting the most relevant type, the revised code ensures that all meaningful output is considered in the height estimation.
Benefits of the Changes
Enhanced Accuracy: The new logic captures a broader range of output types, leading to a more accurate estimation of cell height.
Dynamic Adaptability: The function can now adapt to different types of outputs, making it more robust and flexible for various use cases.
Improved User Experience: With better height estimations, the UI can render cells more effectively, enhancing the overall user experience in Jupyter notebooks.
In summary, these changes significantly improve the functionality of the estimateWidgetSize method, ensuring that it provides a more accurate representation of cell heights based on a wider array of output types.

### CODE REVIEW

File: packages/notebook/src/windowing.ts [hunk 1]
```diff
@@ -61,18 +61,35 @@ export class NotebookViewModel extends WindowedListModel {
...
       }
     }
-    return (
-      NotebookViewModel.DEFAULT_EDITOR_LINE_HEIGHT * (nLines + outputsLines) +
-      NotebookViewModel.DEFAULT_CELL_MARGIN
-    );
+
+    const totalLines = nLines + outputsLines;
+    this.cellsEstimatedHeight.set(model.id, totalLines);
+    return totalLines;
```
Comment (krassowski) [inline packages/notebook/src/windowing.ts:92]: This function ought to return an estimated size in pixels, not the number of lines - can you fix it?

File: packages/notebook/src/windowing.ts [hunk 2]
```diff
@@ -61,18 +61,35 @@ export class NotebookViewModel extends WindowedListModel {
...
         }
       }
     }
-    return (
-      NotebookViewModel.DEFAULT_EDITOR_LINE_HEIGHT * (nLines + outputsLines) +
-      NotebookViewModel.DEFAULT_CELL_MARGIN
-    );
+
+    const totalLines = nLines + outputsLines;
+    this.cellsEstimatedHeight.set(model.id, totalLines);
```
Comment (krassowski) [inline packages/notebook/src/windowing.ts:91]: This function should not set the estimated height as it is not its responsibility.

**Review by krassowski — COMMENTED** [review]
Thanks for starting this! Leaving some initial feedback inline.

File: packages/notebook/src/windowing.ts (hunk 2 above)
Comment (doshi-kevin) [reply packages/notebook/src/windowing.ts:91]: sure thing, I will make the necessary changes and send another pull request. I am just starting open-source and enjoying this, It would be great to be a contributor to Jupyterlabs.

Comment (krassowski) [reply packages/notebook/src/windowing.ts:91]: Usually you do not need to open a new pull request, just push changes to this branch and this PR will get updated :)

Comment (doshi-kevin) [reply packages/notebook/src/windowing.ts:91]: Hey @krassowski , sorry I didn't know about that, I have made the mentioned changes here #16708. Could you please check it and tell me corrections if any.

Comment (doshi-kevin) [reply packages/notebook/src/windowing.ts]: Hey @krassowski any updates on this?
I made a new pull request, could you please check it.

File: packages/notebook/src/windowing.ts [hunk 3]
```diff
@@ -52,27 +52,58 @@ export class NotebookViewModel extends WindowedListModel {
   estimateWidgetSize = (index: number): number => {
     const model = this.cells[index].model;
     const height = this.cellsEstimatedHeight.get(model.id);
+
+
```
Comment (krassowski) [inline packages/notebook/src/windowing.ts:56]:
```suggestion

```

Let's avoid using double new line in side of functions.

File: packages/notebook/src/windowing.ts [hunk 4]
```diff
@@ -52,27 +52,58 @@ export class NotebookViewModel extends WindowedListModel {
...
     if (typeof height === 'number') {
       return height;
     }
 
+
```
Comment (krassowski) [inline packages/notebook/src/windowing.ts:61]:
```suggestion
```

File: packages/notebook/src/windowing.ts [hunk 5]
```diff
@@ -52,27 +52,58 @@ export class NotebookViewModel extends WindowedListModel {
...
         }
       }
     }
-    return (
-      NotebookViewModel.DEFAULT_EDITOR_LINE_HEIGHT * (nLines + outputsLines) +
-      NotebookViewModel.DEFAULT_CELL_MARGIN
-    );
+    const totalLines = nLines + outputsLines;
+
+
```
Comment (krassowski) [inline packages/notebook/src/windowing.ts:102]:
```suggestion
```

**Review by krassowski — COMMENTED** [review]
Thank you! Can you add a test case for this as well, following the existing test in:

https://github.com/jupyterlab/jupyterlab/blob/91f1c7c53c78920920a627267dcbae7e9eafed49/packages/notebook/test/modelfactory.spec.ts#L96-L122

As for the code changes description, it would be ok (or even preferred) to write one-two lines summary (for a PR of this ### size). The intent here is to have a high-level overview and call out anything that may need attention (like changes to public API) - it should allow to get a quick understanding of the PR without looking at the source code.

File: packages/notebook/src/windowing.ts [hunk 6]
```diff
@@ -52,27 +52,58 @@ export class NotebookViewModel extends WindowedListModel {
...
     if (model instanceof CodeCellModel) {
+      const supportedOutputTypes = [
+        'text/html',
+        'image/svg+xml',
+        'application/pdf',
+        'text/markdown',
+        'text/plain',
+        'application/vnd.jupyter.stderr',
+        'application/vnd.jupyter.stdout',
+        'text'
```
Comment (krassowski) [inline packages/notebook/src/windowing.ts:74]:
```suggestion
        'text/markdown',
        'text/plain',
        'application/vnd.jupyter.stderr',
        'application/vnd.jupyter.stdout',
        'text'
```

Maybe we should stick to only these as estimates from html/xml code would be likely unreliable?

File: packages/notebook/src/windowing.ts [hunk 7]
```diff
@@ -52,27 +52,58 @@ export class NotebookViewModel extends WindowedListModel {
...
     if (typeof height === 'number') {
       return height;
     }
 
+
     const nLines = model.sharedModel.getSource().split('\n').length;
     let outputsLines = 0;
+
     if (model instanceof CodeCellModel) {
+      const supportedOutputTypes = [
```
[the remark this answers is not in the document]
Comment (krassowski) [reply packages/notebook/src/windowing.ts:66]: Or maybe not - if we decide to use a more tailored approach as suggested below, then we would not want to sync it

File: packages/notebook/src/windowing.ts (hunk 6 above)
Comment (doshi-kevin) [reply packages/notebook/src/windowing.ts:74]: @krassowski  on it

**Comment by doshi-kevin** [comment]
Hey @krassowski I have added the test case and removed extra spaces. Kindly check for the same .

File: packages/notebook/src/windowing.ts [hunk 8]
```diff
@@ -52,29 +52,91 @@ export class NotebookViewModel extends WindowedListModel {
...
+    const estimatedHeight = totalLines * lineHeight;
+
+    return estimatedHeight;
   };
 
+  // Add the test case below
+}
+
+// Test case for NotebookViewModel
+describe('NotebookViewModel', () => {
```
Comment (krassowski) [inline packages/notebook/src/windowing.ts:107]: The test case needs to go to the file I referenced earlier

**Comment by doshi-kevin** [comment]
@krassowski  Hey, sorry for inconviniences, I have added that peice of code to modelfactory files and removed extra spaces and edited the code as well.

File: packages/notebook/test/modelfactory.spec.ts [hunk 9]
```diff
@@ -152,7 +152,47 @@ describe('@jupyterlab/notebook', () => {
       const height = notebook.estimateWidgetSize(0);
       expect(height).toBe(124); // (2 source_line + 4 output_line) * 17 (line_height) + 22 (cell_margin)
     });
-  });
+
+    test('should calculate height based on number of lines in source and multiple outputs', () => {
```
Comment (krassowski) [inline packages/notebook/test/modelfactory.spec.ts:156]: This test does not actually test the new code added.

Are you using an LLM model? Which one?

Comment (doshi-kevin) [reply packages/notebook/test/modelfactory.spec.ts:156]: Yeah for the text model I used the perplexity ai llm.

Comment (doshi-kevin) [reply packages/notebook/test/modelfactory.spec.ts:156]: I understood the previous test code and thought that according to the given code added, the test generated by the llm might actually be correct.
Could please tell me where might there be an error so that I can resolve it.

**Review by krassowski — CHANGES_REQUESTED** [review]
Sorry for taking longer to get back here, busy times. `js-notebook` and `lint, 3.12` jobs are failing on CI which indicates that there are problems with code style and the tests not passing - can you fix these?

**Comment by doshi-kevin** [comment]
sure. on it !

**Comment by doshi-kevin** [comment]
Hey @krassowski , I have made some linting changes and changes to the test file. Apart from it, I dont understand why (if it still gives error). I will try my best to understand the issue though...

**Comment by doshi-kevin** [comment]
@krassowski any updates ?

**Comment by doshi-kevin** [comment]
Hey @krassowski , sorry to disturb again, but any updates on this ?

**Comment by JasonWeill** [comment]
@doshi-kevin Your tests are still failing; you can see details inline here: https://github.com/jupyterlab/jupyterlab/pull/16706/files#diff-781a703e5af305446726c378a30e53a445113682c768180c073bc042837e0b14

**Comment by krassowski** [comment]
Closing due to lack of response.
